python/ray/data/_internal/planner/planner.py

```diff
@@ -15,24 +21,32 @@ class Planner:
     done by physical optimizer.
     """
 
-    def plan(self, logical_dag: LogicalOperator) -> PhysicalOperator:
+    def __init__(self):
+        self._physical_op_to_logical_op: Dict[PhysicalOperator, LogicalOperator] = {}
+
+    def plan(self, logical_plan: LogicalPlan) -> PhysicalPlan:
         """Convert logical to physical operators recursively in post-order."""
+        physical_dag = self._plan(logical_plan.dag)
+        return PhysicalPlan(physical_dag, self._physical_op_to_logical_op)
```

Comment:
let's make a defensive copy of `self._physical_op_to_logical_op`, before passing to `PhysicalPlan`?

Reply:
Hmm this map is specifically generated for this `PhysicalPlan`, it shouldn't be used anywhere else, and we are defensively copying in the optimization rule, which I believe is the correct place? https://github.com/ray-project/ray/blob/9afbdcaffdcfa560f08bf6938d9f9fa80d711b44/python/ray/data/_internal/logical/optimizers.py#L53